### ISSUE TITLE
snet: drop spath from snet.Path

### ISSUE THREAD

**Issue by oncilla** [scionproto/scion#4129]
The `spath` package is one of the leftovers from the header V1 to header V2 transition.
It has the baked in assumption that a path is simply a blob of bytes without any additional state.
This is not true for example for the new EPIC path type that NetSec is working on.

I propose the following incomplete changes to `snet.Path` to make it more extendable, and support the EPIC use-case.
In essence, the `snet.Path` interface drops the `Path` field and replaces it with a `Dataplane` field.

The definition of a dataplane path is fairly simple:
```go
// DataplanePath is an abstract representation of a SCION dataplane path.
type DataplanePath interface {
	// SetPath sets the path in the SCION header. It assumes that all the fields
	// except the path and path type are set correctly.
	SetPath(scion *slayers.SCION) error
}
```

In `snet.Packet.Serialize`, we simply invoke the `SetPath` method of the dataplane path without any further operations.
This way, the dataplane path can be a stateful object (which is hidden from snet). This supports the EPIC use-case
and should also support any other upcoming path type.

Changes can be seen here: https://github.com/scionproto/scion/commit/a7b42ae02324696d962474861596c128b12af6eb

**Comment by matzf** [comment]
First, I'd like to restate that I dislike the representation of EPIC-HP as a path type as an extension header would seem to be much more appropriate. See #4099.

The `DataplanePath` interface looks quite neat.
Creating this `DataplanePath` as EPIC path (in the sciond client library) whenever the authenticator info is available seems a bit  at odds with the idea that this would generalize to other path types as well. My suggestion would be to always return the default `DataplanePath` as `SCION` and squirrel away the EPIC authenticator information somewhere in the `PathMetadata` struct. To enable EPIC, we could then replace the dataplane path with the result of some `NewEPICDataplanePath` function.

Side note; maintaining state in the path object generally seems somewhat problematic. Paths need to be periodically refreshed when used for a longer time, and for this we simply replace the path by the new object returned from sciond. In this case just reset the counter, which is unproblematic here because the timestamp is easily unique enough for the packet rates that we can achieve, but it's something that we should take into account.

Side note 2: with the `spath` name being freed up, it would perhaps make sense to rename the `snet/path` subpackage to `spath` (it's a good name after all).

**Comment by oncilla** [comment]
> First, I'd like to restate that I dislike the representation of EPIC-HP as a path type as an extension header would seem to be much more appropriate. See #4099.

Requiring a hop-by-hop extension that needs to process the SCION header itself will be fun. Layer violations etc.
Anyway, I feel like this is not really a discussion that is related to the discussion here. Even if we would have EPIC as an extension, we still want to get rid of spath.

> The DataplanePath interface looks quite neat.
Creating this DataplanePath as EPIC path (in the sciond client library) whenever the authenticator info is available seems a bit at odds with the idea that this would generalize to other path types as well. My suggestion would be to always return the default DataplanePath as SCION and squirrel away the EPIC authenticator information somewhere in the PathMetadata struct. To enable EPIC, we could then replace the dataplane path with the result of some NewEPICDataplanePath function.

Right. I think we can discuss the way to do EPIC path further. I convoluted two parts here, getting rid of spath and showing a way to support EPIC paths. I think no matter the way we decide to do EPIC paths, we want to have this Dataplane path interface.

> Side note; maintaining state in the path object generally seems somewhat problematic. Paths need to be periodically refreshed when used for a longer time, and for this we simply replace the path by the new object returned from sciond. In this case just reset the counter, which is unproblematic here because the timestamp is easily unique enough for the packet rates that we can achieve, but it's something that we should take into account.

Agreed, I think changing the path should actually force a new connection, not simply switching out the dataplane internals.

**Comment by juagargi** [comment]
This proposal should also work well with COLIBRI

**Comment by oncilla** [comment]
Has been implemented in the meantime in #4136
